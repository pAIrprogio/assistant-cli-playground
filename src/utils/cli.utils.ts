import readline from "readline";

function deleteLastLinesFromTerminal(count: number) {
  if (count === 0) return;

  for (let i = 0; i < count; i++) {
    readline.clearLine(process.stdout, 0);
    readline.moveCursor(process.stdout, 0, -1);
  }
  readline.cursorTo(process.stdout, 0);
}

export function deleteLastTextFromTerminal(text: string) {
  const lines = text.split("\n");
  deleteLastLinesFromTerminal(lines.length - 1);
}
