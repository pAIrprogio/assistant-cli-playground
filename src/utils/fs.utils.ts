import { promises as fs } from "fs";
import { join } from "path";

export function readFile(path: string, cwd?: string) {
  const fullPath = cwd ? join(cwd, path) : path;
  return fs.readFile(fullPath, "utf-8").catch((e) => {
    if (e.code === "ENOENT") throw new Error("File does not exist");
    throw e;
  });
}
