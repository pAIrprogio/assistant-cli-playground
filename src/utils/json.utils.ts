export const safeParseJson = (value: string | any): any => {
  if (typeof value === "string") {
    return JSON.parse(value);
  }
  return value;
};

export const objectEntriesToJson = (object: Record<string, any>) => {
  return Object.fromEntries(
    Object.entries(object).map(([key, value]) => [key, JSON.stringify(value)]),
  );
};

export const objectEntriesFromJson = (object: Record<string, string>) => {
  return Object.fromEntries(
    Object.entries(object).map(([key, value]) => [key, safeParseJson(value)]),
  );
};
