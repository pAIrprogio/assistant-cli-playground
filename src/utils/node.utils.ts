import { fileURLToPath } from "url";
import { dirname, join } from "path";

export const toDirname = (url: string) => dirname(fileURLToPath(url));

export const ROOTDIR = join(toDirname(import.meta.url), "../..");

export const waitFor = (ms: number) =>
  new Promise((resolve) => setTimeout(resolve, ms));

export const retry = async <T>(fn: () => Promise<T>, retries: number) => {
  for (let attempt = 1; attempt <= retries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      if (attempt === retries) {
        throw error; // Rethrow error if last attempt fails
      }
      return await new Promise((resolve) => setTimeout(resolve, 1000)); // Wait 1 second before retrying
    }
  }
};
