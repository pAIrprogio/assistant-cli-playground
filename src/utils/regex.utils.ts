export const regexFilter = (regex?: string | RegExp) => {
  if (!regex) return (input: string) => true;
  if (regex instanceof RegExp) return (input: string) => regex.test(input);

  const filterRegex = new RegExp(regex);
  return (input: string) => filterRegex.test(input);
};
