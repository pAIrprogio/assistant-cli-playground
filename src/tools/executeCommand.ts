import { z } from "zod";
import { $ } from "zx";
import { Tool } from "../tool.utils.js";

const argsSchema = z.object({
  command: z.string(),
});

type Args = z.input<typeof argsSchema>;

export const executeCommand = {
  name: "executeCommand",
  description: "Executes a command in a bash terminal",
  argsSchema,
  async call(args: Args) {
    // Split space but keep quoted strings together
    const argsArray = args.command.match(/"[^"]+"|\S+/g) || [];
    const res = await $`${argsArray}`;

    return {
      success: true,
      output: res.stdout,
    };
  },
} satisfies Tool<Args>;
