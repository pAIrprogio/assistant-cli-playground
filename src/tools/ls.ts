import { z, input } from "zod";
import { $ } from "zx";
import { Tool } from "../tool.utils.js";

const argsSchema = z.object({
  relativePath: z
    .string()
    .default(".")
    .describe("relative path to the directory"),
});

type Args = z.input<typeof argsSchema>;

export const ls = {
  name: "ls",
  description: "List a directory's files",
  argsSchema,
  async call({ relativePath }: Args) {
    const res =
      await $`(git ls-files ${relativePath}; git ls-files -m  ${relativePath}; git ls-files --others --exclude-standard  ${relativePath}) | sort | uniq`;
    return {
      success: true,
      output: res.stdout.trim(),
    };
  },
} satisfies Tool<Args>;
