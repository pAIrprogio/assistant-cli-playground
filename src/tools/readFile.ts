import { readFile as fsReadFile } from "fs/promises";
import { z } from "zod";
import { Tool } from "../openai/tool.utils.js";

const argsSchema = z.object({
  relativeFilePath: z
    .string()
    .describe("The path to the file relative to the project root"),
  prefixWithLineNumbers: z
    .boolean()
    .optional()
    .describe("Prefix each line with its line number. Use it for git patches"),
});

type Args = z.input<typeof argsSchema>;

export default {
  name: "readFile",
  description: "Reads a file",
  argsSchema,
  async call(args: Args) {
    let fileContent = await fsReadFile(args.relativeFilePath, "utf-8");

    if (args.prefixWithLineNumbers) {
      fileContent = fileContent
        .split("\n")
        .map((line, index) => `${index + 1}:${line}`)
        .join("\n");
    }

    return {
      success: true,
      output: fileContent,
    };
  },
} satisfies Tool<Args>;
