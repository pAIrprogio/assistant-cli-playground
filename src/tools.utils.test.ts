import { describe, expect, it } from "bun:test";
import { toOpenAiTool, toOpenAiTools } from "./tool.utils.js";
import { ls } from "./tools/ls.js";
import { readFile } from "./tools/readFile.js";
import { executeCommand } from "./tools/executeCommand.js";

describe("toOpenAiTool", () => {
  it("creates schema from a tool", () => {
    expect(toOpenAiTool(ls)).toMatchSnapshot();
  });
});

describe("toOpenAiTools", () => {
  it("creates schemas from tools", () => {
    expect(toOpenAiTools([readFile, executeCommand])).toMatchSnapshot();
  });
});
