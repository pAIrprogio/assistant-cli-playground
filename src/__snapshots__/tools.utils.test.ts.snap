// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toOpenAiTool > creates schema from a tool 1`] = `
{
  "function": {
    "description": "List a directory's files",
    "name": "ls",
    "parameters": {
      "additionalProperties": false,
      "properties": {
        "relativePath": {
          "default": ".",
          "description": "relative path to the directory",
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "type": "function",
}
`;

exports[`toOpenAiTools > creates schemas from tools 1`] = `
[
  {
    "function": {
      "description": "Reads a file",
      "name": "readFile",
      "parameters": {
        "additionalProperties": false,
        "properties": {
          "prefixWithLineNumbers": {
            "description": "Prefix each line with its line number. Use it for git patches",
            "type": "boolean",
          },
          "relativeFilePath": {
            "description": "The path to the file relative to the project root",
            "type": "string",
          },
        },
        "required": [
          "relativeFilePath",
        ],
        "type": "object",
      },
    },
    "type": "function",
  },
  {
    "function": {
      "description": "Executes a command in a bash terminal",
      "name": "executeCommand",
      "parameters": {
        "additionalProperties": false,
        "properties": {
          "command": {
            "type": "string",
          },
        },
        "required": [
          "command",
        ],
        "type": "object",
      },
    },
    "type": "function",
  },
]
`;
